src/safeposix/syscalls/fs_calls.rs

```diff
@@ -1898,87 +1898,126 @@ impl Cage {
-                            let fcntlret;
-                            if arg & O_NONBLOCK == O_NONBLOCK {
-                                //set for non-blocking I/O
-                                fcntlret = ins.set_nonblocking();
-                            } else {
-                                //clear non-blocking I/O
-                                fcntlret = ins.set_blocking();
-                            }
-                            if fcntlret < 0 {
-                                match Errno::from_discriminant(interface::get_errno()) {
-                                    Ok(i) => {
-                                        return syscall_error(
-                                            i,
-                                            "fcntl",
-                                            "The libc call to fcntl failed!",
-                                        );
+        //if the provided file descriptor is out of bounds, get_filedescriptor returns Err(),
+        //matching on which throws a 'Bad file number' error
+        //otherwise, file descriptor table entry is stored in 'checkedfd'
+        let fd_table_entry_ptr = self.get_filedescriptor(fd);
+        match fd_table_entry_ptr {
+            Err(()) => {
+                syscall_error(Errno::EBADF, "fcntl", "File descriptor is out of range")
+            },
+            Ok(checkedfd) => {
+                let mut unlocked_fd = checkedfd.write();
+                //returning a 'Bad file number' error if the file descriptor entry is empty
+                //performing the specified command otherwise
+                let fd_table_entry = &mut *unlocked_fd;
+                match fd_table_entry {
+                    None => syscall_error(Errno::EBADF, "fcntl", "Invalid file descriptor"),
+                    Some(filedesc_enum) => {
+                        //'flags' consists of bitwise-or'd access mode, file creation, and file status flags
+                        //to retrieve a particular flag, it can bitwise-and'd with 'flags'
+                        let flags = match filedesc_enum {
+                            Epoll(obj) => &mut obj.flags,
+                            Pipe(obj) => &mut obj.flags,
+                            Stream(obj) => &mut obj.flags,
+                            File(obj) => &mut obj.flags,
+                            //not clear why running F_SETFL on Socket type requires special treatment
+                            Socket(ref mut sockfdobj) => {
+                                if cmd == F_SETFL && arg >= 0 {
+                                    let sock_tmp = sockfdobj.handle.clone();
+                                    let mut sockhandle = sock_tmp.write();
+
+                                    if let Some(ins) = &mut sockhandle.innersocket {
+                                        let fcntlret;
+                                        if arg & O_NONBLOCK == O_NONBLOCK {
```

Comment:
Deep nesting.  Is there a way we can avoid this?

Reply:
No not easily. The rustfmt makes it looks worse than it is.

Reply:
I think clippy will suggest a way to change this pattern:
```
       match fd_table_entry_ptr {
            Err(()) => {
                syscall_error(Errno::EBADF, "fcntl", "File descriptor is out of range")
            },
            Ok(checkedfd) => {
                let mut unlocked_fd = checkedfd.write();
                ...
```
into an `if let Ok(checkedfd) = fd_table_entry_ptr {...}` in at least some cases.  Here, it would unfortunately move the error condition to the end, so while this would reduce indentation, it isn't a 100% win...

Reply:
This seems fine as it for now but may be part of a later refactor.